contrib/migration/migrate-image-manifests.sh

```diff
@@ -0,0 +1,238 @@
+#!/bin/bash
+
+set -iuo pipefail
+IFS=$'\n\t'
+
+readonly AUDITOR_ROLE='system:image-auditor'
+
+readonly USAGE="Usage: $(basename ${BASH_SOURCE[0]}) [OPTIONS]
+
+It fetches manifests of images from OpenShift integrated registry forcing it to
+store the manifests on the local storage and remove them from etcd. This will
+work as long as PR https://github.com/openshift/origin/pull/11925 is compiled
+in the regisry binary.
+
+The script doesn't make any changes unless -a flag is given.
+
+See bug https://bugzilla.redhat.com/show_bug.cgi?id=1378180 for more details.
+
+Options:
+  -h                 Print this message and exit.
+  -a                 Apply. Do the changes.
+  -r <registry_url>  Specify registry url to connect to.
+                     If not given, script will try to determine from
+                     cluster status.
+  -s                 Registry is secured.
+  -c <cacert>        CA certificate for use with registry if the registry is
+                     secured.
+  -t <token>         Token to use to query the registry.
+                     Its user or service account must be able to
+                     get imagestreams/layers in all the namespaces.
+                     If not given, token of current user will be used.
+                     Run this command as cluster admin to give particular
+                     user enough rights to query the images:
+
+                       \$ oadm policy add-cluster-role-to-user registry-viewer <user>
+
+  -f                 Force migration of externally managed images (those imported
+                     from remote registries). A migration attempt is done by
+                     default. If it fails (manifest cannot be stored on
+                     registry's storage because of missing dependencies), the
+                     manifest will be kept in image object (and in etcd). This
+                     option causes a removal of the manifest in any way. It
+                     will still be available only on the remote registry. If
+                     pullthrough feature is enabled, registry will serve such
+                     manifests from the external locations.
+
+                     For this to work, the user must be an image auditor:
+
+                       \$ oadm policy add-cluster-role-to-user ${AUDITOR_ROLE} <user>
+"
+
+registry_address=""
+secured=0
+cacert=""
+token=""
+force_removal_of_manifests=0
+dry_run=1
+
+function get_docker_registry_url() {
+    local tmpl=$'{{.spec.clusterIP}}%{{range $i, $port := .spec.ports}}{{$port.targetPort}}@{{$port.port}},{{end}}'
+    IFS=% read -r service_ip ports <<<"$(oc get -o go-template="${tmpl}" -n default svc/docker-registry)"
```

Comment:
I know that this function is called only in the subshell, but
```
local service_ip ports
```
to prevent the leaking of those variables outside the function.